File: web_ui/src/lib/display.js
/** Pure display helpers for the Web SPA (unit-testable). */

const PRESENCE_SORT_ORDER = { available: 0, away: 1, busy: 2, offline: 3 };

export function filterContacts(contacts, query) {
  const q = (query || '').trim().toLowerCase();
  if (!q) return contacts;
  return contacts.filter(
    (c) =>
      c.name?.toLowerCase().includes(q) ||
      c.jid?.toLowerCase().includes(q) ||
      c.id?.toLowerCase().includes(q),
  );
}

export function findLocalContact(contacts, localJid) {
  if (!localJid) return null;
  const needle = localJid.toLowerCase();
  return contacts.find((c) => c.jid?.toLowerCase() === needle) || null;
}

export function isTransportConnected(connection) {
  if (!connection?.transports?.length) return false;
  return connection.transports.some((t) => t.state === 'connected');
}

export function isAwaitingConnection(connection) {
  return !isTransportConnected(connection);
}

export function formatLocalIdentity(localJid, localContact = null) {
  if (!localJid) return 'You: …';
  if (localContact) {
    return `You: ${localContact.name} (${localJid})`;
  }
  return `You: ${localJid} · not listed in address book`;
}

export function formatHashCompact(contentHash, shortLen = 12) {
  if (!contentHash) return '';
  const short = contentHash.replace('SHA256:', '').slice(0, shortLen);
  return `Book ${short}…`;
}

export function sortContacts(contacts, presence, mode = 'status') {
  const list = [...contacts];
  if (mode === 'name') {
    return list.sort((a, b) => (a.name || '').localeCompare(b.name || '', undefined, { sensitivity: 'base' }));
  }
  return list.sort((a, b) => {
    const aShow = presence[a.id]?.show || 'offline';
    const bShow = presence[b.id]?.show || 'offline';
    const orderDiff = (PRESENCE_SORT_ORDER[aShow] ?? 99) - (PRESENCE_SORT_ORDER[bShow] ?? 99);
    if (orderDiff !== 0) return orderDiff;
    return (a.name || '').localeCompare(b.name || '', undefined, { sensitivity: 'base' });
  });
}

export function prepareContactList(contacts, presence, { needle = '', sortMode = 'status' } = {}) {
  const filtered = filterContacts(contacts, needle);
  return sortContacts(filtered, presence, sortMode);
}

export function presenceColor(show) {
  if (show === 'available') return 'text-green-400';
  if (show === 'away') return 'text-yellow-400';
  if (show === 'busy') return 'text-orange-400';
  return 'text-slate-500';
}

export function formatTime(ts) {
  if (!ts) return '';
  try {
    return new Date(ts).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
  } catch {
    return String(ts).slice(0, 5);
  }
}

export function formatSyncStatusLine(syncStatus) {
  if (!syncStatus) return '';
  if (!syncStatus.secret_configured) {
    return 'Not configured — set addressbook.sync.secret in config';
  }
  const state = syncStatus.enabled ? 'Enabled' : 'Disabled';
  const mode = syncStatus.auto_apply ? 'auto-apply' : 'manual review';
  return `${state} · ${mode} · ${syncStatus.pending_count ?? 0} pending`;
}

export function formatPendingUpdate(update) {
  if (!update) return '';
  return `${update.action || '?'} from ${update.from_jid || '?'} (${update.contact_id || '?'})`;
}

export function buildStatusLine(transports = [], contactCount = 0, outbox = 0) {
  const base = transports.map((t) => `${t.transport}:${t.state}`).join(' · ') || 'unknown';
  let line = `${base} · ${contactCount} contacts`;
  if (outbox > 0) line += ` · outbox ${outbox}`;
  return line;
}

export function connectionStatusFromError(error, connected) {
  if (connected) return 'connected';
  if (!error) return 'connecting';
  return `error: ${error.message || error}`;
}

export function hashCells(contentHash = '', hashBlocks = [], grid = 8) {
  const hex = contentHash.startsWith('SHA256:') ? contentHash.slice(7) : contentHash;
  const palette = [
    '#1e1e2e', '#45475a', '#585b70', '#6c7086',
    '#f38ba8', '#fab387', '#f9e2af', '#a6e3a1',
    '#94e2d5', '#89b4fa', '#cba6f7', '#f5c2e7',
    '#b4befe', '#74c7ec', '#89dceb', '#a6adc8',
  ];
  if (hashBlocks.length >= grid * grid) {
    return hashBlocks.slice(0, grid * grid);
  }
  return hex
    .split('')
    .map((ch) => palette[parseInt(ch, 16)] || '#45475a')
    .slice(0, grid * grid);
}

File: src/xmpp_p2p_chat/text_ui/display.py
"""Pure display helpers for the Textual TUI (unit-testable)."""

from __future__ import annotations

from datetime import datetime

from xmpp_p2p_chat.common.addressbook_hash import render_hash_grid_rich


def presence_symbol(show: str) -> str:
    return {"available": "●", "away": "◐", "busy": "◉", "offline": "○"}.get(show, "○")


def transport_label(contact: dict) -> str:
    transport = contact.get("preferred_transport", "xmpp-server")
    if transport == "direct-p2p":
        direct = contact.get("direct") or {}
        return f"P2P {direct.get('host', '?')}:{direct.get('port', '?')}"
    server = contact.get("xmpp_server") or "auto"
    return f"XMPP ({server})"


def format_contact_detail(contact: dict, presence: dict) -> str:
    pres = presence.get(contact.get("id", ""), {})
    show = pres.get("show", "offline")
    status = pres.get("status", "")
    lines = [
        f"[bold]{contact.get('name', '?')}[/]  ({contact.get('id', '?')})",
        f"JID: [cyan]{contact.get('jid', '?')}[/]",
        f"Transport: {transport_label(contact)}",
        f"Presence: {presence_symbol(show)} {show}" + (f" — {status}" if status else ""),
    ]
    tags = contact.get("tags") or []
    if tags:
        lines.append(f"Tags: {', '.join(tags)}")
    notes = contact.get("notes") or ""
    if notes:
        lines.append(f"Notes: {notes}")
    direct = contact.get("direct")
    if direct:
        fp = direct.get("public_key_fingerprint")
        lines.append(f"Direct: {direct.get('host')}:{direct.get('port')}")
        if fp:
            lines.append(f"Fingerprint: [dim]{fp[:48]}…[/]" if len(fp) > 48 else f"Fingerprint: {fp}")
    return "\n".join(lines)


def format_contact_summary(
    *,
    version: int | str,
    contact_count: int,
    online_count: int,
    warning_count: int = 0,
) -> str:
    warn = f" · [yellow]{warning_count} warn[/]" if warning_count else ""
    return f"v{version} · {contact_count} contacts · {online_count} online{warn}"


PRESENCE_SORT_ORDER = {"available": 0, "away": 1, "busy": 2, "offline": 3}


def find_local_contact(contacts: list[dict], local_jid: str) -> dict | None:
    needle = local_jid.lower()
    for contact in contacts:
        if contact.get("jid", "").lower() == needle:
            return contact
    return None


def is_transport_connected(connection: dict | None) -> bool:
    if not connection:
        return False
    return any(t.get("state") == "connected" for t in connection.get("transports") or [])


def format_local_identity(local_jid: str, local_contact: dict | None = None) -> str:
    if local_contact:
        name = local_contact.get("name", local_jid)
        return f"[bold]You:[/] {name} [dim]({local_jid})[/]"
    return f"[bold]You:[/] [cyan]{local_jid}[/] [dim]· not listed in address book[/]"


def format_hash_compact(content_hash: str, *, short_len: int = 12) -> str:
    if not content_hash:
        return ""
    short = content_hash.replace("SHA256:", "")[:short_len]
    return f"[dim]Book {short}…[/]"


def format_hash_awaiting(content_hash: str, *, grid: int = 8) -> str:
    if not content_hash:
        return "[yellow]Awaiting connection[/] — loading address book…"
    short = content_hash.replace("SHA256:", "")[:12]
    return (
        f"[yellow]Awaiting connection[/] — verify book hash [dim]{short}…[/]\n"
        + render_hash_grid_rich(content_hash, grid=grid)
    )


def sort_contacts(
    contacts: list[dict],
    presence: dict[str, dict],
    *,
    mode: str = "status",
) -> list[dict]:
    if mode == "name":
        return sorted(contacts, key=lambda c: c.get("name", "").lower())
    return sorted(
        contacts,
        key=lambda c: (
            PRESENCE_SORT_ORDER.get(presence.get(c["id"], {}).get("show", "offline"), 99),
            c.get("name", "").lower(),
        ),
    )


def prepare_contact_list(
    contacts: list[dict],
    presence: dict[str, dict],
    *,
    needle: str = "",
    sort_mode: str = "status",
) -> list[dict]:
    filtered = filter_contacts(contacts, needle)
    return sort_contacts(filtered, presence, mode=sort_mode)


def format_hash_sidebar(content_hash: str, *, grid: int = 8, short_len: int = 16) -> str:
    """Full hash grid for settings / address book screens."""
    if not content_hash:
        return ""
    short = content_hash.replace("SHA256:", "")[:short_len]
    return f"[dim]Book hash {short}…[/]\n" + render_hash_grid_rich(content_hash, grid=grid)


def format_message_timestamp(ts: str | datetime) -> str:
    if isinstance(ts, datetime):
        return ts.strftime("%H:%M")
    if isinstance(ts, str) and "T" in ts:
        try:
            return datetime.fromisoformat(ts.replace("Z", "+00:00")).strftime("%H:%M")
        except ValueError:
            return ts[:5] if len(ts) >= 5 else ts
    return str(ts)[:5] if ts else ""


def format_message_line(
    msg: dict,
    *,
    peer_name: str = "Peer",
) -> tuple[str, str]:
    """Return (style_class, rich_text_line) for one chat message."""
    ts = format_message_timestamp(msg.get("timestamp", ""))
    direction = msg.get("direction", "in")
    name = "You" if direction == "out" else peer_name
    style = "msg-out" if direction == "out" else "msg-in"
    status = msg.get("status", "")
    marker = " [green]✓[/]" if status in ("delivered", "sent") else (
        " [yellow]…[/]" if status == "pending" else ""
    )
    align = "right" if direction == "out" else "left"
    body = msg.get("body", "")
    line = (
        f"[{style}][{align}][dim]{ts}[/dim] {name}:[/{align}]\n"
        f"[{style}][{align}]{body}{marker}[/{align}][/{style}]"
    )
    return style, line


def format_message_log(messages: list[dict], *, peer_name: str = "Peer") -> str:
    if not messages:
        return "[dim]No messages yet. Say hello![/]"
    return "\n\n".join(format_message_line(m, peer_name=peer_name)[1] for m in messages)


def format_pending_update_line(update: dict) -> str:
    action = update.get("action", "?")
    from_jid = update.get("from_jid", "?")
    contact_id = update.get("contact_id", "?")
    return f"[cyan]{action}[/] from {from_jid} · contact [bold]{contact_id}[/]"


def format_sync_status_panel(
    *,
    sync_status: dict,
    pending_updates: list[dict] | None = None,
    max_pending_lines: int = 4,
) -> str:
    enabled = sync_status.get("enabled", False)
    auto_apply = sync_status.get("auto_apply", False)
    pending_count = sync_status.get("pending_count", 0)
    secret_configured = sync_status.get("secret_configured", False)

    if not secret_configured:
        state = "[yellow]Not configured[/] — set [addressbook.sync] secret in config"
    elif enabled:
        state = "[green]Enabled[/]"
    else:
        state = "[dim]Disabled[/]"

    mode = "auto-apply" if auto_apply else "manual review"
    lines = [
        f"[bold]Address book sync[/]  {state}",
        f"Mode: {mode} · Pending: {pending_count}",
    ]

    pending = pending_updates or []
    if pending_count and not pending:
        lines.append("[dim]Loading pending updates…[/]")
    elif pending:
        for update in pending[:max_pending_lines]:
            lines.append(f"  • {format_pending_update_line(update)}")
        if len(pending) > max_pending_lines:
            lines.append(f"  [dim]…and {len(pending) - max_pending_lines} more[/]")

    return "\n".join(lines)


def format_addressbook_status_panel(
    *,
    health: dict,
    ab_status: dict,
    connection: dict,
    contact_count: int | None = None,
) -> str:
    ok = health.get("ok", True)
    count = contact_count if contact_count is not None else ab_status.get("contact_count", 0)
    version = ab_status.get("version", 0)
    warnings = (ab_status.get("warnings") or []) + (health.get("warnings") or [])
    outbox = health.get("pending_outbox", 0)
    uptime = int(health.get("uptime_seconds", 0))
    transports = connection.get("transports") or []
    transport_line = " · ".join(
        f"{t.get('transport', '?')}:{t.get('state', '?')}" for t in transports
    ) or "no transports"
    fp = connection.get("p2p_fingerprint")
    fp_line = ""
    if fp:
        fp_line = f"\nP2P fingerprint: [dim]{fp[:44]}…[/]" if len(fp) > 44 else f"\nP2P fingerprint: {fp}"

    status_icon = "[green]✓ OK[/]" if ok and not warnings else "[yellow]⚠ Issues[/]"
    warn_block = ""
    if warnings:
        warn_block = "\n[yellow]" + "\n".join(f"• {w}" for w in warnings[:4]) + "[/]"
        if len(warnings) > 4:
            warn_block += f"\n[dim]…and {len(warnings) - 4} more warnings[/]"

    primary = ab_status.get("primary_path", "")
    path_line = f"\n[dim]File: {primary}[/]" if primary else ""

    return (
        f"{status_icon}  [bold]v{version}[/] · {count} contacts · outbox {outbox} · uptime {uptime}s\n"
        f"Transports: {transport_line}{fp_line}{path_line}{warn_block}"
    )


def filter_contacts(contacts: list[dict], needle: str) -> list[dict]:
    q = needle.lower()
    if not q:
        return list(contacts)
    return [
        c
        for c in contacts
        if q in c.get("name", "").lower()
        or q in c.get("jid", "").lower()
        or q in c.get("id", "").lower()
    ]


__all__ = [
    "filter_contacts",
    "find_local_contact",
    "format_addressbook_status_panel",
    "format_contact_detail",
    "format_contact_summary",
    "format_hash_awaiting",
    "format_hash_compact",
    "format_hash_sidebar",
    "format_local_identity",
    "format_message_line",
    "format_message_log",
    "format_message_timestamp",
    "format_pending_update_line",
    "format_sync_status_panel",
    "is_transport_connected",
    "prepare_contact_list",
    "presence_symbol",
    "sort_contacts",
    "transport_label",
]
